src/analyses/relationAnalysis.ml

```diff
@@ -0,0 +1,453 @@
+      let eval_int ctx exp =
+        match ctx.ask (Queries.EvalInt exp) with
+        | x when Queries.ID.is_int x -> Queries.ID.to_int x
+        | _ -> None
+      in
+      let int_min, int_max = IntDomain.Size.range ik in
+      match IntOps.BigIntOps.to_int64 int_min, IntOps.BigIntOps.to_int64 int_max with
+      | exception Failure _ -> false
+      | int_min, int_max ->
+        let min = eval_int ctx (BinOp (Lt, kinteger64 ik int_min, exp , intType)) in
+        let max = eval_int ctx (BinOp (Lt,  exp, kinteger64 ik int_max, intType)) in
+        match min, max with
+        | Some (x), _ when x = (IntOps.BigIntOps.of_int 1) -> true
+        | _, Some (x) when x = (IntOps.BigIntOps.of_int 1) -> true
+        | _, _ -> false
+
+
+  let assert_type_bounds apr x ctx =
+    assert (AD.varinfo_tracked x);
+    let ik = Cilfacade.get_ikind x.vtype in
+    if not (IntDomain.should_ignore_overflow ik) then ( (* don't add type bounds for signed when assume_none *)
+      let (type_min, type_max) = IntDomain.Size.range ik in
+      (* TODO: don't go through CIL exp? *)
+      let e1 = (BinOp (Le, Lval (Cil.var x), (Cil.kintegerCilint ik (Cilint.cilint_of_big_int type_max)), intType)) in
+      let e2 = (BinOp (Ge, Lval (Cil.var x), (Cil.kintegerCilint ik (Cilint.cilint_of_big_int type_min)), intType)) in
+      let apr = AD.assert_inv apr e1 false (no_overflow ctx e1) in
+      let apr = AD.assert_inv apr e2 false (no_overflow ctx e2) in
+      apr
+    )
+    else
+      apr
+
+
+  (* Basic transfer functions. *)
+
+  let assign ctx (lv:lval) e =
+    let st = ctx.local in
+    if !GU.global_initialization && e = MyCFG.unknown_exp then
+      st (* ignore extern inits because there's no body before assign, so the apron env is empty... *)
+    else (
+      if M.tracing then M.traceli "apron" "assign %a = %a\n" d_lval lv d_exp e;
+      let ask = Analyses.ask_of_ctx ctx in
+      let r = assign_to_global_wrapper ask ctx.global ctx.sideg st lv (fun st v ->
+          assign_from_globals_wrapper ask ctx.global st e (fun apr' e' ->
+              AD.assign_exp apr' (RV.local v) e' (no_overflow ctx e')
+            )
+        )
+      in
+      if M.tracing then M.traceu "apron" "-> %a\n" D.pretty r;
+      r
+    )
+
+  let branch ctx e b =
+    let st = ctx.local in
+    let res = assign_from_globals_wrapper (Analyses.ask_of_ctx ctx) ctx.global st e (fun apr' e' ->
+        (* not an assign, but must remove g#in-s still *)
+        AD.assert_inv apr' e' (not b) (no_overflow ctx e')
+      )
+    in
+    if AD.is_bot_env res then raise Deadcode;
+    {st with apr = res}
+
+
+  (* Function call transfer functions. *)
+
+  let enter ctx r f args =
+    let st = ctx.local in
+    if M.tracing then M.tracel "combine" "apron enter f: %a\n" d_varinfo f.svar;
+    if M.tracing then M.tracel "combine" "apron enter formals: %a\n" (d_list "," d_varinfo) f.sformals;
+    if M.tracing then M.tracel "combine" "apron enter local: %a\n" D.pretty ctx.local;
```

Comment:
I guess all these "apron XXX" tracing outputs should be replaced with `relation` or something similar.